File: NativeRuby/ffi.rb
require 'ffi'

module FFIC
  extend FFI::Library
  
 
  ffi_lib './c/main.so'
  
  attach_function:hello,[],:void
  attach_function:sum,[:int,:int],:int
  
end

module FFICPP
  extend FFI::Library
  
 
  ffi_lib './cpp/main.so'
  
  attach_function:hello,[],:void
  attach_function:sum,[:int,:int],:int
  
end

module FFIGO
  extend FFI::Library
  
 
  ffi_lib './go/main.so'
  
  attach_function:Hello,[],:void
  attach_function:Sum,[:int,:int],:int
  
end

module FFIRUST
  extend FFI::Library
  
   ffi_lib './rust/src/librust.so'
  
  attach_function:Hello,[],:void
  attach_function:Sum,[:int,:int],:int
  
end


puts FFIC.sum(1,2)
puts FFIC.hello()

puts FFICPP.sum(3,4)
puts FFICPP.hello()

puts FFIGO.Sum(5,6)
puts FFIGO.Hello()

puts FFIRUST.Sum(7,8)
puts FFIRUST.Hello()

File: NativeJavascript/ffi.js
const ffi = require('ffi');

const c_lib = ffi.Library('./c/main.so', {
    hello: ['void', []],
    sum: ['int', ['int', 'int']]
});
const cpp_lib = ffi.Library('./cpp/main.so', {
    hello: ['void', []],
    sum: ['int', ['int', 'int']]
});
const go_lib = ffi.Library('./go/main.so', {
    Hello: ['void', []],
    Sum: ['int', ['int', 'int']]
});
const rust_lib = ffi.Library('./rust/src/librust.so', {
    Hello: ['void', []],
    Sum: ['int', ['int', 'int']]
});



console.log('\n\njavascript tests')

// c
c_lib.hello()
console.log(c_lib.sum(1, 2))

// cpp
cpp_lib.hello()
console.log(cpp_lib.sum(3, 4))

// go
go_lib.Hello()
console.log(go_lib.Sum(5, 6))

//rust
rust_lib.Hello()
console.log(rust_lib.Sum(7, 8))